fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -1914,13 +1995,61 @@ protected void saveStoreHeader(@Nonnull RecordMetaDataProto.DataStoreInfo storeH
         });
     }
 
+    /**
+     * Set whether the store state is cacheable. In particular, this flag determines whether the
+     * {@linkplain FDBRecordContext#getMetaDataVersionStampAsync(IsolationLevel) meta-data version-stamp} key
+     * is changed whenever the {@link RecordStoreState} is changed. By default, record store state information is
+     * <em>not</em> cacheable. This is because for deployments in which there are many record stores sharing the same
+     * cluster, updating the meta-data version key every time any store changes its state could lead to performance
+     * problems, so (at least for now), this is a feature the user must opt-in to on each record store.
+     *
+     * @param cacheable whether the meta-data version-stamp should be invalidated upon store state change
+     * @return a future that will complete to {@code true} if the store state's cacheability has changed
+     */
+    @Nonnull
+    public CompletableFuture<Boolean> setStateCacheabilityAsync(boolean cacheable) {
+        if (recordStoreStateRef.get() == null) {
+            return preloadRecordStoreStateAsync().thenCompose(vignore -> setStateCacheabilityAsync(cacheable));
+        }
+        if (formatVersion < CACHEABLE_STATE_FORMAT_VERSION) {
+            throw new RecordCoreException("cannot mark record store state cacheable at format version " + formatVersion);
+        }
+        if (recordStoreStateRef.get().getStoreHeader().getCacheable() == cacheable) {
+            return AsyncUtil.READY_FALSE;
+        } else {
+            return updateStoreHeaderAsync(oldHeader -> oldHeader.toBuilder()
+                    .setCacheable(cacheable)
+                    .setLastUpdateTime(System.currentTimeMillis())
+                    .build()
+            ).thenApply(ignore -> true);
+        }
+    }
+
+    /**
+     * Set whether the store state is cacheable. This operation might block if the record store state has
+     * not yet been loaded. Use {@link #setStateCacheabilityAsync(boolean)} in asychronous contexts.
+     *
+     * @param cacheable whether this store's state should be cacheable
+     * @return whether the record store state cacheability has changed
+     * @see #setStateCacheabilityAsync(boolean)
+     */
+    public boolean setStateCacheability(boolean cacheable) {
+        return context.asyncToSync(FDBStoreTimer.Waits.WAIT_SET_STATE_CACHEABILITY, setStateCacheabilityAsync(cacheable));
+    }
+
     // Actually (1) writes the index state to the database and (2) updates the cached state with the new state
     private void updateIndexState(@Nonnull String indexName, byte[] indexKey, @Nonnull IndexState indexState) {
         // This is generally called by someone who should already have a write lock, but adding them here
```

Comment:
It is probably my bad, but now that you are on it, can you please move the comments to the line we take the lock?

Reply:
Done